File: endianness/src/endianness.c
#include <stdio.h>
#include "endianness.h"

int main(int argc, char**argv)
{
    int x1 = 0x01020304;
    short int x2 = 0x0102;

    unsigned char* p1 = (unsigned char*) &x1; 
    unsigned char* p2 = (unsigned char*) &x2;

    int i;
    printf("0x");
    for(i=0; i<sizeof(x1); i++)
        printf("%x", p1[i]);
    printf("\n0x");
    for(i=0; i<sizeof(x2); i++)
        printf("%x", p2[i]);

    x1 = htonl(x1);
    x2 = htons(x2);

    printf("\n0x");
    for(i=0; i<sizeof(x1); i++)
        printf("%x", p1[i]);
    printf("\n0x");
    for(i=0; i<sizeof(x2); i++)
        printf("%x", p2[i]);
    printf("\n");

    return 0;
}

int is_little_endian()
{
    int i=1;
	char* p = (char*) &i;

	return ((int) *p) == i;
}

short int htons(short int num)
{
    int size = sizeof(num);
    short int num2 = 0;
    int i;
    unsigned char* p1 = (unsigned char*) &num;
    unsigned char* p2 = (unsigned char*) &num2;

    if(is_little_endian)
    {
        for(i=0; i<size; i++)
            p2[i]=p1[size-i-1];

        return num2;
    }
    else
        return num;
}

int htonl(int num)
{
    int size = sizeof(num);
    int num2 = 0;
    int i;
    unsigned char* p1 = (unsigned char*) &num;
    unsigned char* p2 = (unsigned char*) &num2;

    if(is_little_endian)
    {
        for(i=0; i<size; i++)
            p2[i]=p1[size-i-1];

        return num2;
    }
    else
        return num;
}

File: endianness/src/Endianness.java
import java.nio.ByteOrder;
import java.nio.ByteBuffer;

public class Endianness
{
    private boolean little_endian;

    public Endianness()
    {
        little_endian = isLittleEndian();
    }

    public boolean isLittleEndian()
    {
        return ByteOrder.nativeOrder().equals(ByteOrder.LITTLE_ENDIAN);
    }
    
    public short htons(short num)
    {
        if(little_endian)
        {
            ByteBuffer bbuf = ByteBuffer.allocate(2);
            bbuf.order(ByteOrder.LITTLE_ENDIAN);
            bbuf.putShort(num);
            bbuf.order(ByteOrder.BIG_ENDIAN);
        
            return bbuf.getShort(0);
        }
        else
            return num;
    }

    public int htonl(int num)
    {
        if(little_endian)
        {
            ByteBuffer bbuf = ByteBuffer.allocate(4);
            bbuf.order(ByteOrder.LITTLE_ENDIAN);
            bbuf.putInt(num);
            bbuf.order(ByteOrder.BIG_ENDIAN);
        
            return bbuf.getInt(0);
        }
        else
            return num;
    }

    public static void main(String argv[])
    {
        Endianness e = new Endianness();
        
        int n1 = 0x01020304;
        short n2 = 0x0102;
        System.out.println("0x0"+Integer.toString(n1,16));
        System.out.println("0x0"+Integer.toString(n2,16));
        
        n1 = e.htonl(n1);
        n2 = e.htons(n2);
        System.out.println("0x0"+Integer.toString(n1,16));
        System.out.println("0x0"+Integer.toString(n2,16));
    }
}
